resources/aws_ec2.go

```diff
@@ -1038,6 +1075,91 @@ func (obj *AwsEc2Res) snsPostHandler(w http.ResponseWriter, req *http.Request) {
 	}
 }
 
+// snsVerifySignature verifies that the post messages are genuine and originate
+// from amazon by checking if the signature is valid for the provided key and
+// message contents.
+func (obj *AwsEc2Res) snsVerifySignature(post postData) error {
+	// download and parse the signing certificate
+	cert, err := obj.snsGetCert(post.SigningCertURL)
+	if err != nil {
+		return errwrap.Wrapf(err, "error getting certificate")
+	}
+	// convert the message to canonical form
+	message := obj.snsCanonicalFormat(post)
+	// decode the message signature from base64
+	signature, err := base64.StdEncoding.DecodeString(post.Signature)
+	if err != nil {
+		return errwrap.Wrapf(err, "error decoding string")
+	}
+	// check the signature against the message
+	err = cert.CheckSignature(x509.SHA1WithRSA, message, signature)
+	if err != nil {
+		return errwrap.Wrapf(err, "error checking signature")
+	}
+	return nil
+}
+
+// snsGetCert downloads and parses the signing certificate from the provided
+// URL for message verification.
+func (obj *AwsEc2Res) snsGetCert(url string) (*x509.Certificate, error) {
+	// only download valid certificates from amazon
+	matchURL, err := regexp.MatchString(SnsCertURLRegex, url)
+	if err != nil {
+		return nil, errwrap.Wrapf(err, "error matching regex")
+	}
+	if !matchURL {
+		return nil, fmt.Errorf("invalid certificate url: %s", url)
+	}
+	// download the signing certificate
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, errwrap.Wrapf(err, "http get error")
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errwrap.Wrapf(err, "error reading post body")
+	}
+	// decode and the certificate and discard the rest
+	decodedCert, _ := pem.Decode(body)
```

Comment:
what's the second arg? an error?

Reply:
It's any junk data that may exist in the response after the pem certificate. I'll make that more clear in the comment.